# The astype Method I

In [2]:
import pandas as pd
nba = pd.read_csv('nba.csv').dropna(how='all')
nba.tail()

Name       Team  Number Position   Age Height  Weight   College  \
452    Trey Lyles  Utah Jazz    41.0       PF  20.0   6-10   234.0  Kentucky   
453  Shelvin Mack  Utah Jazz     8.0       PG  26.0    6-3   203.0    Butler   
454     Raul Neto  Utah Jazz    25.0       PG  24.0    6-1   179.0       NaN   
455  Tibor Pleiss  Utah Jazz    21.0        C  26.0    7-3   256.0       NaN   
456   Jeff Withey  Utah Jazz    24.0        C  26.0    7-0   231.0    Kansas   

        Salary  
452  2239800.0  
453  2433333.0  
454   900000.0  
455  2900000.0  
456   947276.0

In [5]:
nba['Age'].hasnans

False

In [6]:
nba["Age"].astype(dtype='int')

0      25
1      25
2      27
3      22
4      29
       ..
452    20
453    26
454    24
455    26
456    26
Name: Age, Length: 457, dtype: int32

In [7]:
nba['Age'].astype(int)

0      25
1      25
2      27
3      22
4      29
       ..
452    20
453    26
454    24
455    26
456    26
Name: Age, Length: 457, dtype: int32

In [11]:
nba['Age'] = nba['Age'].astype(dtype='int')

In [13]:
nba.tail()

Name       Team  Number Position  Age Height  Weight   College  \
452    Trey Lyles  Utah Jazz    41.0       PF   20   6-10   234.0  Kentucky   
453  Shelvin Mack  Utah Jazz     8.0       PG   26    6-3   203.0    Butler   
454     Raul Neto  Utah Jazz    25.0       PG   24    6-1   179.0       NaN   
455  Tibor Pleiss  Utah Jazz    21.0        C   26    7-3   256.0       NaN   
456   Jeff Withey  Utah Jazz    24.0        C   26    7-0   231.0    Kansas   

        Salary  
452  2239800.0  
453  2433333.0  
454   900000.0  
455  2900000.0  
456   947276.0

In [14]:
nba['Salary']

0      7730337.0
1      6796117.0
2            NaN
3      1148640.0
4      5000000.0
         ...    
452    2239800.0
453    2433333.0
454     900000.0
455    2900000.0
456     947276.0
Name: Salary, Length: 457, dtype: float64

In [15]:
nba['Salary'].hasnans

True

In [16]:
nba['Salary'].astype('int')

IntCastingNaNError: Cannot convert non-finite values (NA or inf) to integer

In [17]:
nba['Salary'] = nba['Salary'].fillna(0)

In [18]:
nba['Salary'].astype(int)

0      7730337
1      6796117
2            0
3      1148640
4      5000000
        ...   
452    2239800
453    2433333
454     900000
455    2900000
456     947276
Name: Salary, Length: 457, dtype: int32

In [22]:
nba['Salary'].hasnans

False

In [23]:
nba

Name            Team  Number Position  Age Height  Weight  \
0    Avery Bradley  Boston Celtics     0.0       PG   25    6-2   180.0   
1      Jae Crowder  Boston Celtics    99.0       SF   25    6-6   235.0   
2     John Holland  Boston Celtics    30.0       SG   27    6-5   205.0   
3      R.J. Hunter  Boston Celtics    28.0       SG   22    6-5   185.0   
4    Jonas Jerebko  Boston Celtics     8.0       PF   29   6-10   231.0   
..             ...             ...     ...      ...  ...    ...     ...   
452     Trey Lyles       Utah Jazz    41.0       PF   20   6-10   234.0   
453   Shelvin Mack       Utah Jazz     8.0       PG   26    6-3   203.0   
454      Raul Neto       Utah Jazz    25.0       PG   24    6-1   179.0   
455   Tibor Pleiss       Utah Jazz    21.0        C   26    7-3   256.0   
456    Jeff Withey       Utah Jazz    24.0        C   26    7-0   231.0   

               College     Salary  
0                Texas  7730337.0  
1            Marquette  6796117.0  
2    Boston University        0.0  
3        Georgia State  1148640.0  
4                  NaN  5000000.0  
..                 ...        ...  
452           Kentucky  2239800.0  
453             Butler  2433333.0  
454                NaN   900000.0  
455                NaN  2900000.0  
456             Kansas   947276.0  

[457 rows x 9 columns]

In [24]:
nba['Salary'] = nba['Salary'].astype(dtype='int')

In [25]:
nba

Name            Team  Number Position  Age Height  Weight  \
0    Avery Bradley  Boston Celtics     0.0       PG   25    6-2   180.0   
1      Jae Crowder  Boston Celtics    99.0       SF   25    6-6   235.0   
2     John Holland  Boston Celtics    30.0       SG   27    6-5   205.0   
3      R.J. Hunter  Boston Celtics    28.0       SG   22    6-5   185.0   
4    Jonas Jerebko  Boston Celtics     8.0       PF   29   6-10   231.0   
..             ...             ...     ...      ...  ...    ...     ...   
452     Trey Lyles       Utah Jazz    41.0       PF   20   6-10   234.0   
453   Shelvin Mack       Utah Jazz     8.0       PG   26    6-3   203.0   
454      Raul Neto       Utah Jazz    25.0       PG   24    6-1   179.0   
455   Tibor Pleiss       Utah Jazz    21.0        C   26    7-3   256.0   
456    Jeff Withey       Utah Jazz    24.0        C   26    7-0   231.0   

               College   Salary  
0                Texas  7730337  
1            Marquette  6796117  
2    Boston University        0  
3        Georgia State  1148640  
4                  NaN  5000000  
..                 ...      ...  
452           Kentucky  2239800  
453             Butler  2433333  
454                NaN   900000  
455                NaN  2900000  
456             Kansas   947276  

[457 rows x 9 columns]

# The astype Method II

In [28]:
nba = pd.read_csv('nba.csv').dropna(how='all')
nba.tail(n=2)

Name       Team  Number Position   Age Height  Weight College  \
455  Tibor Pleiss  Utah Jazz    21.0        C  26.0    7-3   256.0     NaN   
456   Jeff Withey  Utah Jazz    24.0        C  26.0    7-0   231.0  Kansas   

        Salary  
455  2900000.0  
456   947276.0

In [30]:
nba.dtypes

Name         object
Team         object
Number      float64
Position     object
Age         float64
Height       object
Weight      float64
College      object
Salary      float64
dtype: object

In [32]:
nba['Position']

0      PG
1      SF
2      SG
3      SG
4      PF
       ..
452    PF
453    PG
454    PG
455     C
456     C
Name: Position, Length: 457, dtype: object

In [34]:
len(nba)

457

In [35]:
nba['Position'].nunique()

5

In [37]:
nba['Position'].value_counts()

SG    102
PF    100
PG     92
SF     85
C      78
Name: Position, dtype: int64

In [38]:
nba.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 457 entries, 0 to 456
Data columns (total 9 columns):
 #   Column    Non-Null Count  Dtype  
---  ------    --------------  -----  
 0   Name      457 non-null    object 
 1   Team      457 non-null    object 
 2   Number    457 non-null    float64
 3   Position  457 non-null    object 
 4   Age       457 non-null    float64
 5   Height    457 non-null    object 
 6   Weight    457 non-null    float64
 7   College   373 non-null    object 
 8   Salary    446 non-null    float64
dtypes: float64(4), object(5)
memory usage: 35.7+ KB


In [39]:
nba['Position'].info()

<class 'pandas.core.series.Series'>
Int64Index: 457 entries, 0 to 456
Series name: Position
Non-Null Count  Dtype 
--------------  ----- 
457 non-null    object
dtypes: object(1)
memory usage: 7.1+ KB


In [40]:
nba['College'].info()

<class 'pandas.core.series.Series'>
Int64Index: 457 entries, 0 to 456
Series name: College
Non-Null Count  Dtype 
--------------  ----- 
373 non-null    object
dtypes: object(1)
memory usage: 7.1+ KB


In [41]:
nba['Position'] = nba['Position'].astype('category')

In [42]:
nba.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 457 entries, 0 to 456
Data columns (total 9 columns):
 #   Column    Non-Null Count  Dtype   
---  ------    --------------  -----   
 0   Name      457 non-null    object  
 1   Team      457 non-null    object  
 2   Number    457 non-null    float64 
 3   Position  457 non-null    category
 4   Age       457 non-null    float64 
 5   Height    457 non-null    object  
 6   Weight    457 non-null    float64 
 7   College   373 non-null    object  
 8   Salary    446 non-null    float64 
dtypes: category(1), float64(4), object(4)
memory usage: 32.8+ KB


In [43]:
nba['Position'].info()

<class 'pandas.core.series.Series'>
Int64Index: 457 entries, 0 to 456
Series name: Position
Non-Null Count  Dtype   
--------------  -----   
457 non-null    category
dtypes: category(1)
memory usage: 4.2 KB


Name: Position, dtype: int64
<class 'pandas.core.frame.DataFrame'>
Int64Index: 457 entries, 0 to 456
Data columns (total 9 columns):
    Column    Non-Null Count  Dtype  
---  ------    --------------  -----  
 0.   Name      457 non-null    object 
 1.   Team      457 non-null    object 
 2.   Number    457 non-null    float64
 3.   Position  457 non-null    object 
 4.   Age       457 non-null    float64
 5.   Height    457 non-null    object 
 6.   Weight    457 non-null    float64
 7.   College   373 non-null    object 
 8.   Salary    446 non-null    float64
 ***
dtypes: float64(4), object(5)
***
**memory usage: 35.7+ KB**


<class 'pandas.core.series.Series'>
Int64Index: 457 entries, 0 to 456
Series name: Position
Non-Null Count  Dtype 
***
457 non-null    object
dtypes: object(1)
**memory usage: 7.1+ KB**

In [44]:
nba['Team'].nunique()

30

In [45]:
nba['Team'] = nba['Team'].astype('category')

In [46]:
nba.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 457 entries, 0 to 456
Data columns (total 9 columns):
 #   Column    Non-Null Count  Dtype   
---  ------    --------------  -----   
 0   Name      457 non-null    object  
 1   Team      457 non-null    category
 2   Number    457 non-null    float64 
 3   Position  457 non-null    category
 4   Age       457 non-null    float64 
 5   Height    457 non-null    object  
 6   Weight    457 non-null    float64 
 7   College   373 non-null    object  
 8   Salary    446 non-null    float64 
dtypes: category(2), float64(4), object(3)
memory usage: 30.9+ KB


**memory usage: 30.9+ KB**

In [55]:
f'{round((1 - 30.9 / 35.7) * 100,2)}%'

'13.45%'